// functional-json-schema/dist/__tests__/e2e.js
"use strict";
exports.__esModule = true;
var index_1 = require("../lib/index");
var ajv = require("ajv");
describe('types', function () {
    test('sample', function () {
        var generated = index_1.schema({
            community_list: {
                portfolio_ids: index_1.types.arrayOf('string', { required: true }),
                project_id: index_1.types.type('string', { required: true }),
                status: index_1.types.enumOf('string'),
                user: index_1.types.anyOf(index_1.types.definition('User'), index_1.types.definition('Admin'))
            }
        });
        expect(generated).toEqual({
            type: 'object',
            properties: {
                community_list: {
                    type: 'object',
                    properties: {
                        portfolio_ids: {
                            'type': 'array',
                            'items': {
                                'type': 'string'
                            }
                        },
                        project_id: {
                            'type': 'string'
                        },
                        user: {
                            anyOf: [
                                { $ref: '#/definitions/User' },
                                { $ref: '#/definitions/Admin' }
                            ]
                        },
                        status: {
                            "enum": [
                                { type: 'string' }
                            ]
                        }
                    },
                    required: ['portfolio_ids', 'project_id']
                }
            }
        });
        var validator = new ajv();
        expect(validator.validateSchema(generated)).toBe(true);
    });
});
